src/composed/key/mod.rs

```diff
@@ -41,13 +41,64 @@
 //!
 //! [Packet based signing and verifying]: super::super::packet
 //! [signing and verifying with external hashing]: super::signed_key
+//!
+//! # Loading a public key and listing its details
+//!
+//! ```rust
+//! use pgp::composed::parse_public_keys;
+//! use pgp::types::KeyTrait;
+//! use std::path::Path;
+//!
+//! let mut path = Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/debian-10-archive-key.asc");
+//!
+//! let keys = parse_public_keys(&std::fs::read(&path).unwrap()).unwrap();
+//!
+//! for key in &keys {
+//!     println!(
+//!         "key {:?} {:?} {:?}",
+//!         key.key_id(),
+//!         key.fingerprint(),
+//!         &key.details
+//!     );
+//!
+//!     for subkey in &key.public_subkeys {
+//!         println!("subkey {:?} {:?}", subkey.key_id(), subkey.fingerprint());
+//!     }
+//! }
+//! ```
 
 mod builder;
 mod public;
 mod secret;
 mod shared;
 
+use std::io::Cursor;
+
+use crate::{Deserializable, SignedPublicKey};
+
 pub use self::builder::*;
 pub use self::public::*;
 pub use self::secret::*;
 pub use self::shared::*;
+
+/// Parse one or more public keys from binary or ascii armored format.
+///
+/// If any part of the data failed to parse the whole operation will fail.
+///
+/// This is a convenience wrapper around [`SignedPublicKey::from_bytes_many()`]
+/// and [`SignedPublicKey::from_armor_many()`].
+pub fn parse_public_keys(data: &[u8]) -> Result<Vec<SignedPublicKey>, crate::errors::Error> {
```

Comment:
It might be more consistent API wise to move this to `SignedPublicKey:from_many`, not sure though, thoughts?

Reply:
At that point the different argument type (byte slice vs reader) becomes a bit untidy. Of course that can be solved but I frankly don't love Rust's buffering abstractions and prefer to avoid it if possible, and making this a "convenience wrapper" gives an excuse to just use a slice to peek at the first byte. This is not really an argument against moving it, just noting that moving it puts this back onto the "open questions / todo" list.

Making a single lonely function is a bit strange perhaps, and merging it into `SignedPublicKey` is perhaps the obvious thing to do. In fact I think I'll make that change (when I have time).

The main reason I made it a separate function is because of my frustration at how long it took me to understand how I should be loading keys. If you want to understand, read on, but please don't take this personally, I am describing just my annoyance here, but if I didn't appreciate the work that has gone into rPGP I wouldn't be here.

A lot of the code is parsing and (de)serialization. This code contains multiple layers which wrap each other and are named similarly, traits implemented using macros in other modules, etc. This makes it hard to know how a part works without understanding the whole. Part of this is because not just data but also logic is organized around the data model of the RFC, which users don't (want to) know in this much detail. A user knows they have a public key file but in fact that's not a `PublicKey`, it is a `SignedPublicKey`. And these `from_bytes` functions are too camouflaged for me, they look like every other bit of (de)serialization logic (which as I said is a lot). Is the "composed" module even supposed to be public? It is one of only three modules without a description in the top-level doc page, and its doc page describes it as "Internal key definitions". The only reason I put the function there is because of the existing docs there.

In contrast, `parse_public_keys` stands apart from implementation details and has a "verb"-ish name clearly related to what I want to do, instead of a generic sounding deserialization function on a "noun"-ish struct that a user may not even be sure is the right one. Basically, after the annoyance I went through I wanted to try to understand and address the problem in a deeper way than just adding an example, and this was the simplest way.

Reply:
Thanks for the detailed comments, some notes below

> This makes it hard to know how a part works without understanding the whole. Part of this is because not just data but also logic is organized around the data model of the RFC, which users don't (want to) know in this much detail. 

Yes this is intended for the current level that is implemented, as this was meant as a low level building block to allow folks who want to build their own tooling, or support their own use cases. I always intended to eventually have a high level, user friendly api that did not require reading the whole RFC before using it :D 

Because of that it might make sense to actually take this method and start putting it into its own module, that can eventually be built out to provide such a higher level api. 

> Is the "composed" module even supposed to be public? It is one of only three modules without a description in the top-level doc page, and its doc page describes it as "Internal key definitions". 

It is another low level building block, which certainly needs better documentation, and is mainly exposed currently in this form to make things usable, not because it is a well and nice defined API